modules/nlu-testing/src/tools/visualisation.ts

```diff
@@ -0,0 +1,149 @@
+import * as sdk from 'botpress/sdk'
+import _ from 'lodash'
+import { distance, similarity } from 'ml-distance'
+import { Matrix } from 'ml-matrix'
+import { PCA } from 'ml-pca'
+
+import { BotState } from '../backend/typings'
+
+const clustering = require('density-clustering')
+
+export async function computeEmbeddingSimilarity(state: BotState) {
+  const intentDatas = _.groupBy(state.trainDatas, 'intent')
+
+  const intentEmb = {}
+  for (const key in intentDatas) {
+    if (Object.prototype.hasOwnProperty.call(intentDatas, key)) {
+      const meanEmb = intentDatas[key]
+        .map(o => o.utt_emb)
+        .reduce((tot, cur) => {
+          return Matrix.add(tot, new Matrix([cur]))
```

Comment:
Seems more like a vector add than a matrix add? Why the need for Matrix class?

Reply:
ok nvm, I noticed your doing an element wise division few lines down. I have no opinion on this library.